zipline/utils/calendars/trading_calendar.py

```diff
@@ -0,0 +1,843 @@
+        """
+        idx = next_divider_idx(self.market_closes_nanos, dt.value)
+        return self.schedule.market_close[idx].tz_localize('UTC')
+
+    def previous_open(self, dt):
+        """
+        Given a dt, returns the previous open.
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to get the previous open.
+
+        Returns
+        -------
+        pd.Timestamp
+            The UTC imestamp of the previous open.
+        """
+        idx = previous_divider_idx(self.market_opens_nanos, dt.value)
+        return self.schedule.market_open[idx].tz_localize('UTC')
+
+    def previous_close(self, dt):
+        """
+        Given a dt, returns the previous close.
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to get the previous close.
+
+        Returns
+        -------
+        pd.Timestamp
+            The UTC timestamp of the previous close.
+        """
+        idx = previous_divider_idx(self.market_closes_nanos, dt.value)
+        return self.schedule.market_close[idx].tz_localize('UTC')
+
+    def next_minute(self, dt):
+        """
+        Given a dt, return the next exchange minute.  If the given dt is not
+        an exchange minute, returns the next exchange open.
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to get the next exchange minute.
+
+        Returns
+        -------
+        pd.Timestamp
+            The next exchange minute.
+        """
+        idx = next_divider_idx(self._trading_minutes_nanos, dt.value)
+        return pd.Timestamp(self._trading_minutes_nanos[idx], tz='UTC')
+
+    def previous_minute(self, dt):
+        """
+        Given a dt, return the previous exchange minute.
+
+        Raises KeyError if the given timestamp is not an exchange minute.
+
+        Parameters
+        ----------
+        dt: pd.Timestamp
+            The dt for which to get the previous exchange minute.
+
+        Returns
+        -------
+        pd.Timestamp
+            The previous exchange minute.
+        """
+
+        idx = previous_divider_idx(self._trading_minutes_nanos, dt.value)
+        return pd.Timestamp(self._trading_minutes_nanos[idx], tz='UTC')
+
+    def next_session_label(self, session_label):
+        """
+        Given a session label, returns the label of the next session.
+
+        Parameters
+        ----------
+        session_label: pd.Timestamp
+            A session whose next session is desired.
+
+        Returns
+        -------
+        pd.Timestamp
+            The next session label (midnight UTC).
+
+        Notes
+        -----
+        Raises ValueError if the given session is the last session in this
+        calendar.
+        """
+        idx = self.schedule.index.get_loc(session_label)
+        try:
+            return self.schedule.index[idx + 1]
+        except IndexError:
+            if idx == len(self.schedule.index) - 1:
```

Comment:
Is there a case where `idx != len(self.schedule.index) - 1` where we want to swallow the `IndexError`, as we do here?